url7b90c1010b_DastanIqbal_HelloZMQwithJava_tgz-pJ8-ClientJ8/src/zmq/V1Decoder.java

```diff
@@ -0,0 +1,152 @@
+    private static final int flags_ready = 2;
+    private static final int message_ready = 3;
+    
+    private final byte[] tmpbuf;
+    private Msg in_progress;
+    private IMsgSink msg_sink;
+    private final long maxmsgsize;
+    private int msg_flags;
+    
+    public V1Decoder (int bufsize_, long maxmsgsize_, IMsgSink session)
+    {
+        super (bufsize_);
+
+        maxmsgsize = maxmsgsize_;
+        msg_sink = session;
+        
+        tmpbuf = new byte[8];
+        
+        //  At the beginning, read one byte and go to one_byte_size_ready state.
+        next_step (tmpbuf, 1, flags_ready);
+    }
+
+    //  Set the receiver of decoded messages.
+// JDK6    @Override
+    public void set_msg_sink (IMsgSink msg_sink_) 
+    {
+        msg_sink = msg_sink_;
+    }
+
+    @Override
+    protected boolean next() {
+        switch(state()) {
+        case one_byte_size_ready:
+            return one_byte_size_ready ();
+        case eight_byte_size_ready:
+            return eight_byte_size_ready ();
+        case flags_ready:
+            return flags_ready ();
+        case message_ready:
+            return message_ready ();
+        default:
+            return false;
+        }
+    }
+
+
+
+    private boolean one_byte_size_ready () {
+        
+        int size = tmpbuf [0];
+        if (size < 0)
+            size = (0xff) & size;
+
+        //  Message size must not exceed the maximum allowed size.
+        if (maxmsgsize >= 0)
+            if (size > maxmsgsize) {
+                decoding_error ();
+                return false;
+            }
+
+        //  in_progress is initialised at this point so in theory we should
+        //  close it before calling zmq_msg_init_size, however, it's a 0-byte
+        //  message and thus we can treat it as uninitialised...
+        in_progress = new Msg (size);
+
+        in_progress.set_flags (msg_flags);
+        next_step (in_progress.data (), in_progress.size (),
+            message_ready);
+
+        return true;
+    }
+    
+    private boolean eight_byte_size_ready() {
+        
+        //  The payload size is encoded as 64-bit unsigned integer.
+        //  The most significant byte comes first.
+        final long msg_size = ByteBuffer.wrap(tmpbuf).getLong();
+
+        //  Message size must not exceed the maximum allowed size.
+        if (maxmsgsize >= 0)
+            if (msg_size > maxmsgsize) {
+                decoding_error ();
+                return false;
+            }
+
+        //  Message size must fit within range of size_t data type.
+        if (msg_size > Integer.MAX_VALUE) {
+            decoding_error ();
+            return false;
+        }
+        
+        //  in_progress is initialised at this point so in theory we should
+        //  close it before calling init_size, however, it's a 0-byte
+        //  message and thus we can treat it as uninitialised.
+        in_progress = new Msg ((int) msg_size);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code assigns a resource to a class field and there is no method that closes or returns this resource. This can result in a resource leak that slows down or crashes your system.

**Fix**
Create a method in this class that closes the following resource: *in\_progress*. After you use this resource, use this method to prevent a resource leak.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).